### PR TITLE
Feature/index creation jenkinsfile

### PR DESCRIPTION
**Pull request by ONS-Tom**
- Add `Jenkinsfile` with Checkout, Index Exists and Create Index steps, with a parameter for `index_name`
- Add shell scripts for checking if an index exists and for creating a new index - these have been moved to the `business-index-api`, [here](https://github.com/ONSdigital/business-index-api/tree/feature/elastic-shell-scripts/conf/scripts)

Previously, if the index name provided by the user already existed in ElasticSearch, it would be deleted. After further clarification, it was decided to not delete the index and to fail the script instead.

Environment variables used by the `Jenkinsfile` (ElasticSearch host, port, index name) have been left empty in the `environment` section, they will be added as actual environment variables once other parts of the pipeline have been tested.

### CODE REVIEW

File: scripts/create_index.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+port=$2
+index_name=$3
+path_to_index_definition=$4
+
+# Read in the index definition JSON
+create_index_json=`cat $path_to_index_definition`
+
+# Make the curl PUT request to create the index, passing in our index definition JSON
+echo 'Creating ElasticSearch index...'
+curl -XPUT "$host:$port/$index_name" -d"$create_index_json"
```
Comment (willis7) [inline scripts/create_index.sh]: You could remove L24 and use:

`curl -XPUT "${host}:${port}/${index_name}" -d @"${path_to_index_definition}"`

Note also the style for wrapping expanded variables with braces. Whilst both are valid, I find the style above easier to read when constructing urls.

File: scripts/create_index.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    echo 'Error, you need to provide the following arguments:'
+    echo 'host, port, index name and the path to the index JSON file.'
+    echo 'e.g. sh scripts/create_index.sh localhost 9200 bi-dev-test-index ./json/create_index.json'
+    exit 1
+fi
+
+host=$1
+port=$2
+index_name=$3
+path_to_index_definition=$4
```
Comment (willis7) [inline scripts/create_index.sh]: Constants and anything exported to the environment should be capitalized.

File: scripts/create_index.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# This script will create an ElasticSearch index using passed in command line
+# arguments for the host, port, index name and path to JSON file which includes
+# the index definition.
+
+# Fail fast if we get any errors
+set -e
+
+# Fail the script if we recieve an incorrect number of arguments
+if [[ $# -ne 4 ]] ; then
```
Comment (willis7) [inline scripts/create_index.sh]: `4` is a magic number. Use a variable with a sensible name `REQUIRED_ARGS` etc.

Generally, if I'm taking more than a single argument I will use `getopts`. It makes the script easier to maintain.

File: scripts/create_index.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    echo 'Error, you need to provide the following arguments:'
+    echo 'host, port, index name and the path to the index JSON file.'
+    echo 'e.g. sh scripts/create_index.sh localhost 9200 bi-dev-test-index ./json/create_index.json'
```
Comment (willis7) [inline scripts/create_index.sh]: Traditionally you document your arguments in the usage() function:

```
function usage {
    echo "usage: $programname [-abch] [-f infile] [-o outfile]"
    echo "  -a      turn on feature a"
    echo "  -b      turn on feature b"
    echo "  -c      turn on feature c"
    echo "  -h      display help"
    echo "  -f infile   specify input file infile"
    echo "  -o outfile  specify output file outfile"
    exit 1
}

usage
```

Comment (willis7) [reply scripts/create_index.sh]: The example here is what you may find with getopts

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            agent any
+            when {
+                anyOf {
+                    branch BRANCH_MASTER
+                }
+            }
+            steps {
+                script {
+                    colourText("info", "Checking to see if index [$INDEX_NAME] exists")
+                    sh "chmod 777 './scripts/index_exists.sh'" // Make sure we can execute the script
```
Comment (willis7) [inline Jenkinsfile]: You shouldn't need to do this. It looks like git didn't save the execute bit to the index. This often happens when checking in shell scripts on windows.

You can fix that: `git update-index --chmod=+x $script`

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            when {
+                anyOf {
+                    branch BRANCH_MASTER
+                }
+            }
+            steps {
+                deleteDir()
+                checkout scm
+                dir('business-index-api') {
+                    git(url: "https://github.com/ONSdigital/business-index-api.git", branch: "master")
```
Comment (willis7) [inline Jenkinsfile]: Not sure how I feel about this. If this repo is somewhat coupled to business-index-api, then maybe the files in this repo should live there?

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+pipeline {
+    environment {
+        BRANCH_MASTER = "master"
+        ELASTIC_HOST = ""
+        ELASTIC_PORT = ""
+        INDEX_NAME = ""
+        INDEX_JSON_PATH = "./business-index-api/conf/index.json"
+    }
+    options {
+        skipDefaultCheckout()
```
Comment (willis7) [inline Jenkinsfile]: Why are you skipping default checkout?

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    }
+    agent any
+    stages {
+        stage('Checkout'){
+            agent any
+            when {
+                anyOf {
+                    branch BRANCH_MASTER
+                }
+            }
```
Comment (willis7) [inline Jenkinsfile]: At minimum I would inline this ` when{ anyOf{ branch "master" } }`, but more importantly - why is this needed? If you want this exclusive to master you can simply use `when{ branch "master" }`.

File: Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            steps {
+                script {
+                    colourText("info", "Checking to see if index [$INDEX_NAME] exists")
+                    sh "chmod 777 './scripts/index_exists.sh'" // Make sure we can execute the script
+                    sh "./scripts/index_exists.sh $ELASTIC_HOST $ELASTIC_PORT $INDEX_NAME"
```
Comment (willis7) [inline Jenkinsfile]: Aren't `$ELASTIC_HOST $ELASTIC_PORT $INDEX_NAME"` all set to empty strings?

File: scripts/index_exists.sh [hunk 10]
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
```
Comment (willis7) [inline scripts/index_exists.sh]: This duplicates a large amount of  `create_index.sh`. 

With getopts you could have a single script with different functions that execute dependent on the flags set.

File: scripts/index_exists.sh [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+host=$1
+port=$2
+index_name=$3
+
+echo "Checking if index [$index_name] already exists within ElasticSearch"
+http_code=$(curl --write-out "%{http_code}\n" --silent --output /dev/null "$host:$port/$index_name")
+
+echo "HTTP Code returned from checking if the index [$index_name] exists: $http_code"
+if [[ $http_code == '200' ]] ; then
```
Comment (willis7) [inline scripts/index_exists.sh]: Nice!

**Review by willis7 — CHANGES_REQUESTED** [review]
Good work! I've suggested a few fixes we can discuss.

File: scripts/create_index.sh (hunk 1 above)
Comment (ONS-Tom) [reply scripts/create_index.sh]: That's really useful!

File: Jenkinsfile (hunk 8 above)
Comment (ONS-Tom) [reply Jenkinsfile]: Having `master` set in one place makes it useful for testing, when working off a feature branch so it only needs to be changed in one place. The `anyOf` isn't needed though.

File: Jenkinsfile (hunk 7 above)
Comment (ONS-Tom) [reply Jenkinsfile]: According to Ameen, it's due to the default checkout cloning the repository on each node and not just the master, but I'll try removing it and see what happens.

File: Jenkinsfile (hunk 6 above)
Comment (ONS-Tom) [reply Jenkinsfile]: Good point, do you just mean the scripts or the actual `Jenkinsfile` too?

File: Jenkinsfile (hunk 9 above)
Comment (ONS-Tom) [reply Jenkinsfile]: For now. `ELASTIC_HOST` and `ELASTIC_PORT` will need to be added to Jenkins global variables, although I'll request this later on incase I need anymore added. I need to think about how to either make the `INDEX_NAME` accept user input (might be a temporary solution) or to fully automate the only manual aspect of the current freestyle job (aliasing an index to bi-dev/test/beta after data has been loaded)

Comment (willis7) [reply Jenkinsfile]: Sounds like a job for parameters:
https://jenkins.io/doc/book/pipeline/syntax/#parameters

File: Jenkinsfile (hunk 7 above)
Comment (willis7) [reply Jenkinsfile]: Further down you run `checkout scm` which is the same as default checkout ;-)

You typically only need to disable default checkout if you need special checkout behaviour (such as shallow clones)

File: Jenkinsfile (hunk 8 above)
Comment (willis7) [reply Jenkinsfile]: Fair enough, that can be a PITA.

File: scripts/index_exists.sh (hunk 10 above)
Comment (ONS-Tom) [reply scripts/index_exists.sh]: I think I'll leave it for now and see how many more similar scripts I need to make, if there is more of the same then I'll move them into a single script.

File: Jenkinsfile (hunk 9 above)
Comment (ONS-Tom) [reply Jenkinsfile]: I hadn't realised that parameters can be used with a `Jenkinsfile`!

**Review by willis7 — APPROVED** [review]
:+1:

**Comment by willis7** [comment]
You should probably run an interactive rebase and squash all these commits ;-)
